Improve Headless CMS and Asset Store docs

kazydek opened the pull request:
**Description**

Changes proposed in this pull request:

- Improved Headless CMS and Asset Store docs --> see #5174 for detailed requirements. 
- Moved part of the docs to Service Catalog and Console docs.
- Changed the `UI Contracts` document type into `Console UI Views` to make it more general, as it was extended with info on documentation in the Service Catalog and Instances views
- Improved choice columns in Asset Store and Headless CMS docs.

**Related issue(s)**
See also #5174. It also relates to documentation requirements in https://github.com/kyma-project/kyma/issues/5467.

michal-hudy commented on docs/asset-store/06-03-clusterasset.md:
```diff
@@ -86,43 +65,68 @@ status:
+
+| Reason | Phase | Description |
+| --------- | ------------- | ----------- |
+| `Pulled` | `Ready` | The Asset Controller pulled the asset content for processing. |
```
Same content as for Asset CR, also maybe we can move this section to a separate doc, because right now for both types it always will be the same.

kazydek replied:
I will think about it. I was already considering it but decided that since it is an integral part of the CRD, I will keep it that way - but true, there is no differentiation in the description between ClusterAssets and Assets... @derberg what do you think?

michal-hudy commented on docs/asset-store/06-04-clusterbucket.md:
```diff
@@ -36,21 +36,36 @@ status:
 
+### Status reasons
+
+Processing of a ClusterBucket CR can succeed, continue, or fail for one of these reasons:
```
Same content as for Bucket CR, also maybe we can move this section to a separate doc, because right now for both types it always will be the same.

kazydek replied:
as above

michal-hudy commented on docs/headless-cms/03-02-cms-asyncapi-service.md:
```diff
@@ -0,0 +1,12 @@
+
+CMS AsyncAPI Service is an HTTP server used in Kyma to process AsyncAPI specifications. It only accepts `multipart/form-data` forms and contains two endpoints:
+
+- `/validate` that validates the AsyncAPI specification against the AsyncAPI schema in version 2.0.0.
```
You can mention that it uses https://github.com/asyncapi/parser :)

kazydek replied:
Will do 👍

michal-hudy commented on docs/headless-cms/06-01-docstopic.md:
```diff
@@ -49,26 +49,49 @@ status:
+
+| Reason | Phase | Description |
+| --------- | ------------- | ----------- |
+| `AssetCreated` | `Ready` | The Asset Controller created the specified asset. |
```
Pending, not Asset Controller, CMS - also in other places ;)

kazydek replied:
Changed to DocsTopic Controller here and everywhere else.

michal-hudy commented on docs/headless-cms/06-01-docstopic.md:
```diff
@@ -49,26 +49,49 @@ status:
+| `AssetDeleted` | `Pending` | The Asset Controller deleted specified assets. |
+| `AssetDeletionFailed` | `Failed`  | The Asset Controller couldn't delete the specified asset due to the provided error. |
+| `AssetsDeletionFailed` | `Failed` | The Asset Controller couldn't delete assets due to the provided error. |
+| `AssetUpdated` | `Ready` | The Asset Controller updated the specified asset. |
```
Pending

michal-hudy commented on docs/headless-cms/06-02-clusterdocstopic.md:
```diff
@@ -45,26 +45,49 @@ status:
 
+### Status reasons
+
+Processing of a ClusterDocsTopic CR can succeed, continue, or fail for one of these reasons:
```
Same comments as for non cluster

kazydek commented on docs/asset-store/03-04-supported-webhooks.md:
```diff
@@ -0,0 +1,41 @@
+
+Asset Store supports the following types of webhooks:
+
+- **Mutation webhook** acts similarly to the validation service. The difference is that it mutates the asset instead of just validating it. For example, this can mean asset rewriting through the `regex` operation or `keyvalue`, or the modification in the JSON specification. The mutation webhook returns modified files instead of information on the status.
```
So true, thanks.

kazydek commented on docs/asset-store/03-04-supported-webhooks.md:
```diff
@@ -0,0 +1,41 @@
+  - accept **parameters** and **content** properties.
+  - return `200` and `304` responses.
+
+See the example of the validation service with the `/convert` endpoint [here](./assets/mutation-service.yaml). To preview and test it, use the [Swagger Editor](https://editor.swagger.io).
```
As discussed, there is already such an issue.

kazydek commented on docs/application-connector/01-01-application-connector.md:
```diff
@@ -33,6 +33,6 @@ You can register an API secured with one of the following authentication methods
+You can register any API that adheres to the REST principles and is available over the HTTP protocol. The Application Connector also allows you to register APIs implemented with the OData technology.
 
-You can provide specifications that document your APIs. The Application Connector supports [OpenAPI](https://www.openapis.org/) and [OData](https://www.odata.org/documentation) specification formats.
+You can provide specifications that document your APIs. The Application Connector supports [AsyncAPI](https://www.asyncapi.com/), [OpenAPI](https://www.openapis.org/), and [OData](https://www.odata.org/documentation) specification formats.
```
Done 👍

kazydek commented on docs/asset-store/03-04-supported-webhooks.md:
```diff
@@ -0,0 +1,42 @@
+
+Asset Store supports the following types of webhooks:
+
+- **Mutation webhook** mutates fetched assets. For example, this can mean asset rewriting through the `regex` operation or `keyvalue`, or the modification in the JSON specification. The mutation webhook returns modified files instead of information on the status.
```
sure, why not

kazydek commented on docs/asset-store/03-04-supported-webhooks.md:
```diff
@@ -0,0 +1,42 @@
+
+- **Mutation webhook** mutates fetched assets. For example, this can mean asset rewriting through the `regex` operation or `keyvalue`, or the modification in the JSON specification. The mutation webhook returns modified files instead of information on the status.
+
+- **Validation webhook** validates fetched assets before the Asset Controller uploads them into the bucket. It can be a list of several different validation webhooks and all of them should be processed even if one fails. It can refer either to the validation of a specific file against a specification or to the security validation. The validation webhook returns the validation status when the vali ...
```
ok, I unified them ( chose version 2)

kazydek commented on docs/asset-store/03-04-supported-webhooks.md:
```diff
@@ -0,0 +1,42 @@
+
+- **Mutation webhook** mutates fetched assets. For example, this can mean asset rewriting through the `regex` operation or `keyvalue`, or the modification in the JSON specification. The mutation webhook returns modified files instead of information on the status.
+
+- **Validation webhook** validates fetched assets before the Asset Controller uploads them into the bucket. It can be a list of several different validation webhooks and all of them should be processed even if one fails. It can refer either to the validation of a specific file against a specification or to the security validation. The validation webhook returns the validation status when the vali ...
```
`are`

kazydek commented on docs/asset-store/03-04-supported-webhooks.md:
```diff
@@ -0,0 +1,42 @@
+
+- **Mutation webhook** mutates fetched assets. For example, this can mean asset rewriting through the `regex` operation or `keyvalue`, or the modification in the JSON specification. The mutation webhook returns modified files instead of information on the status.
+
+- **Validation webhook** validates fetched assets before the Asset Controller uploads them into the bucket. It can be a list of several different validation webhooks and all of them should be processed even if one fails. It can refer either to the validation of a specific file against a specification or to the security validation. The validation webhook returns the validation status when the vali ...
```
got it 👌

kazydek commented on docs/asset-store/03-04-supported-webhooks.md:
```diff
@@ -0,0 +1,42 @@
+
+- **Validation webhook** validates fetched assets before the Asset Controller uploads them into the bucket. It can be a list of several different validation webhooks and all of them should be processed even if one fails. It can refer either to the validation of a specific file against a specification or to the security validation. The validation webhook returns the validation status when the vali ...
+
+- **Metadata webhook** allows you to extract metadata from assets and inserts it under the `status.assetRef.files.metadata` field in the (Cluster)Asset CR. For example, the Asset Metadata Service which is the metadata webhook implementation in Kyma, extracts front matter metadata from `.md` files and returns the status with such information as `title` and `type`. The metadata is later used to ren ...
```
Well, I mentioned a particular Kyma use case of Asset Metadata Service but sure, I can remove that.

kazydek commented on docs/asset-store/03-04-supported-webhooks.md:
```diff
@@ -0,0 +1,42 @@
+
+## Service specification requirements
+
+If you create a specific mutation, validation, or metadata service for the available webhooks and you want the Asset Store to properly communicate with it, you must ensure that the API exposed by the given service meets certain format requirements. These criteria differ depending on the service type:
```
`it` meaning with the specific mutation, validation, or metadata **service**

kazydek commented on docs/asset-store/03-04-supported-webhooks.md:
```diff
@@ -0,0 +1,42 @@
+
+## Service specification requirements
+
+If you create a specific mutation, validation, or metadata service for the available webhooks and you want the Asset Store to properly communicate with it, you must ensure that the API exposed by the given service meets certain format requirements. These criteria differ depending on the service type:
```
ok

kazydek commented on docs/asset-store/03-04-supported-webhooks.md:
```diff
@@ -0,0 +1,42 @@
+
+## Service specification requirements
+
+If you create a specific mutation, validation, or metadata service for the available webhooks and you want the Asset Store to properly communicate with it, you must ensure that the API exposed by the given service meets certain format requirements. These criteria differ depending on the service type:
```
Hmm why mutation only? I change it into `webhook type` instead - is that ok?

kazydek commented on docs/asset-store/03-04-supported-webhooks.md:
```diff
@@ -0,0 +1,42 @@
+
+- **metadata service** must expose endpoints that:
+
+  - pass file data in the in `"object": "string"` format in the request body, where **object** stands for the file name and **string** is the file content.
```
Corrected

kazydek commented on docs/console/03-04-headless-cms-in-console.md:
```diff
@@ -20,3 +20,20 @@ To define how DocsTopic and ClusterDocsTopic are rendered in the UI, use the fol
+
+| Type | Description | Format | Version |
+| --------- | ------------- | ----------- | ----------- |
+| [OpenAPI](https://www.openapis.org/) |   API-related information  | `json` and `yaml`| 2.0 and lower |
```
ok, so I changed it into 3.0

kazydek commented on docs/console/03-04-headless-cms-in-console.md:
```diff
@@ -20,3 +20,20 @@ To define how DocsTopic and ClusterDocsTopic are rendered in the UI, use the fol
+| [AsyncAPI](https://www.asyncapi.com/) |   Messaging data (for Events)  | `json` and `yaml`| 2.0 and lower |
+| Markdown |  Service Class or component documentation  | `md`|  |
+
+All API and messaging specifications are converted to the `json` format in the Console UI. However, Kyma uses its own implementation of the validation and mutation webhooks, called CMS AsyncAPI Service, to validate if the AsyncAPI input files are in the `2.0` version. If not, it converts them to `2.0`. The service also changes any AsyncAPI `yaml` input file to the `json` format:
```
As far as I know, yes :) --> [link](https://github.com/kyma-project/kyma/tree/master/components/cms-services/cmd/asyncapi#cms-asyncapi-service)

kazydek commented on docs/console/03-04-headless-cms-in-console.md:
```diff
@@ -20,3 +20,20 @@ To define how DocsTopic and ClusterDocsTopic are rendered in the UI, use the fol
+| [AsyncAPI](https://www.asyncapi.com/) |   Messaging data (for Events)  | `json` and `yaml`| 2.0 and lower |
+| Markdown |  Service Class or component documentation  | `md`|  |
+
+All API and messaging specifications are converted to the `json` format in the Console UI. However, Kyma uses its own implementation of the validation and mutation webhooks, called CMS AsyncAPI Service, to validate if the AsyncAPI input files are in the `2.0` version. If not, it converts them to `2.0`. The service also changes any AsyncAPI `yaml` input file to the `json` format:
```
I changed that part a bit (pls take a look at the related diagram as well when you are going to review them again)

kazydek commented:
@derberg:
-  I added all the changes. 
- Regarding your comment about a missing section in HCMS docs on `webhook-config-map.yaml` ConfigMap, I added this info in the `headless-cms-in-the-console` doc as a tip. I thought it is a good place since I already described there (and presented on the diagram) the default webhooks/services applied in Kyma for the specs. If you have a better idea where we can place that info, let me know. 
- Also, let me know if it is ok to leave the descriptions of status resources for assets, buckets, and DocsTopics in the CRDs or if you prefer to move them to separate docs.

mmitoraj commented on docs/service-catalog/11-04-specifications-in-console-ui.md:
```diff
@@ -14,9 +42,9 @@ another_metadata: {value or text string}
+### Metadata
 
-Each Markdown document displayed in the Kyma Console requires metadata in a specific format called [Front Matter](https://forestry.io/docs/editing/front-matter/).
+Each Markdown document displayed in the Kyma Console UI requires metadata in a specific format called [Front Matter](https://forestry.io/docs/editing/front-matter/).
```
```suggestion
Each Markdown document displayed in the Kyma Console UI requires metadata in a specific format called [front matter](https://forestry.io/docs/editing/front-matter/).
```

michal-hudy commented on docs/asset-store/06-01-asset.md:
```diff
@@ -88,44 +66,70 @@ status:
 
 > **NOTE:** The Asset Controller automatically adds all parameters marked as **Not applicable** to the Asset CR.
 
+### Status reasons
```
Yes, it is possible, we will improve that ;)

michal-hudy commented on docs/asset-store/03-04-supported-webhooks.md:
```diff
@@ -0,0 +1,42 @@
+- **validation service** must expose endpoints that:
+
+  - contain **parameters** and **content** properties.
+  - return the `200` response with new file content.
```
Validation service doesn't return file content.
 - 200 - means OK
 - 422 - validation failed - and message with errors

michal-hudy commented on docs/asset-store/03-04-supported-webhooks.md:
```diff
@@ -0,0 +1,42 @@
+- **metadata service** must expose endpoints that:
+
+  - pass file data in the `"object": "string"` format in the request body, where **object** stands for the file name and **string** is the file content.
+  - return the `200` response with new file content.
```
Not file content, extracted metadata

kazydek commented on docs/asset-store/06-01-asset.md:
```diff
@@ -88,44 +66,70 @@ status:
+| `BucketNotReady` | `Pending` | The referenced bucket is not ready. |
+| `BucketError` | `Failed` | Reading the bucket status failed due to the provided error. |
+| `Mutated` | `Pending` | Mutation services changed the asset content. |
+| `MutationFailed` | `Failed` | Asset mutation failed for one of the provided reasons. |
```
Exactly :)

kazydek commented on docs/console/03-04-headless-cms-in-console.md:
```diff
@@ -20,3 +20,25 @@ To define how DocsTopic and ClusterDocsTopic are rendered in the UI, use the fol
 | Parameter | Default Value | Description |
 | --------- | ------------- | ----------- |
 | **spec.sources.metadata.disableRelativeLinks** | `false` | Disables relative links when documentation is rendered. It only applies to the `markdown` type of assets included in the (Cluster)DocsTopic CR. |
+
```
changed into `ServiceClass documentation`
